# File:     04-pop_stats.snake
# Name:     Neal Platt
# Date:     13 April 2018
# Desc:     This Snake file is used to generate sH vcfs with sM coordinates and
#           calculate basic popgen stats including struture plots, pca, fst,etc.

# Reccomended
# Usage:    snakemake \
#               --printshellcmds \
#               --use-conda \
#               --cluster 'qsub -V -cwd -S /bin/bash -pe mpi {threads} -o {log}.log -e {log}.stderr' \
#               --jobs 100 \
#               --latency-wait 200 \
#               --snake \
#               --keep-going
#               --restart-times 2

# Reqs:     All required software should be installed/available through the 
#           singularity image, packaged conda env, or in the scripts dir.

# Comments: This portion of the workflow exectures after refining the genotypes

#-------------------------------------------------------------------------------
# To Do:
#   3] update comments
#   4] add cleanup on success
#   5] would prefer to run structure rather than admixture.
#   6] Add R code for building plots
#   7] mitochondrial PCA
#-------------------------------------------------------------------------------

#setting up snakemake 
HOME_DIR = "/master/nplatt/schisto_hybridization/"

CONTIG_CHUNKS=50
ASSEMBLY = "schHae_v1.fa"

#setting work directory
RESULTS_DIR = HOME_DIR + "results/"
DATA_DIR = HOME_DIR + "data/"
LOG_DIR = RESULTS_DIR + "logs/"
GENOME_DIR = DATA_DIR + "genome/"
LISTS_DIR = RESULTS_DIR + "lists/"

GENOME = GENOME_DIR + ASSEMBLY
SMAN_GENOME = GENOME_DIR + "Smansoni_v7.fa"
SCUR_GENOME = GENOME_DIR + "GCA_000951415.1_S_curassoni_Dakar_genomic.fna"
SBOV_GENOME = GENOME_DIR + "Schistosoma_bovis_v1.0.fa"

SEQ_DIR = DATA_DIR + "sequence_data/"
SCRIPTS_DIR = HOME_DIR + "scripts/"
CONFIG_DIR = HOME_DIR + "config/"
SINGULARITY_IMG = CONFIG_DIR + "snpCalling_v0.0.8.img"

PROCESS_READS_DIR = RESULTS_DIR + "01-processed_reads/"
FILTER_READS_DIR = PROCESS_READS_DIR + "filtered_reads/"
MAP_DIR = PROCESS_READS_DIR + "sh_mapped_reads/"

GATK_DIR = RESULTS_DIR + "02-sh_bqsr_geno/" 
BQSR_R1_DIR = GATK_DIR + "r1/"
R1_HC = BQSR_R1_DIR + "hc/"
R1_DB = BQSR_R1_DIR + "db/"
R1_GENO = BQSR_R1_DIR + "geno/"
R1_MERGE = BQSR_R1_DIR + "merge/"
R1_FILTER = BQSR_R1_DIR + "filter/"
R1_BAM = BQSR_R1_DIR + "bams/"
R1_TABLE = BQSR_R1_DIR + "tables/"

BQSR_R2_DIR = GATK_DIR + "r2/"
R2_HC = BQSR_R2_DIR + "hc/"
R2_DB = BQSR_R2_DIR + "db/"
R2_GENO = BQSR_R2_DIR + "geno/"
R2_MERGE = BQSR_R2_DIR + "merge/"
R2_FILTER = BQSR_R2_DIR + "filter/"
R2_BAM = BQSR_R2_DIR + "bams/"
R2_TABLE = BQSR_R2_DIR + "tables/"

REFPAN_DIR = RESULTS_DIR + "03-build_refpanel/"
 
OUTGENO_DIR = RESULTS_DIR + "04-genotype_outgroups/"
OUT_MAP = OUTGENO_DIR + "mapped_reads/"
OUT_BQ_1_HC = OUTGENO_DIR + "bqsr_1_hc/"
OUT_BQ_1_DB = OUTGENO_DIR + "bqsr_1_db/"
OUT_BQ_1_GENO = OUTGENO_DIR + "bqsr_1_geno/"
OUT_BQ_1_MERGE = OUTGENO_DIR + "bqsr_1_merge/"
OUT_BQ_1_FILTER = OUTGENO_DIR + "bqsr_1_filter/"
OUT_BQ_2_HC = OUTGENO_DIR + "bqsr_2_hc/"
OUT_BQ_2_DB = OUTGENO_DIR + "bqsr_2_db/"
OUT_BQ_2_GENO = OUTGENO_DIR + "bqsr_2_geno/"
OUT_BQ_2_MERGE = OUTGENO_DIR + "bqsr_2_merge/"
OUT_BQ_2_FILTER = OUTGENO_DIR + "bqsr_2_filter/"

WGA_DIR = RESULTS_DIR + "06-wga/" 


HAE_SAMPLES = [ "Sh.NE_Dai-002.1",      "Sh.NE_Dai-010.1",      "Sh.NE_Dai-013.3",      
            	"Sh.NE_Dai-031.1",      "Sh.NE_Dai-033.1",      "Sh.NE_Dai-044.1",
            	"Sh.NE_Dai-045.1",      "Sh.NE_Dai-051.1",      "Sh.NE_Dai-074.1",      
            	"Sh.NE_Dai-146.1",      "Sh.NE_DaiCP-233.1",    "Sh.NE_DaiCP-276.1", 
           	    "Sh.NE_DaiCP-277.2",    "Sh.NE_Doki-029.1",     "Sh.NE_Kar-001.1",      
            	"Sh.NE_Kar-002.1",      "Sh.NE_Kar-076.1",      "Sh.NE_Kar-096.2",      
            	"Sh.NE_Kar-241.1",      "Sh.NE_Kar-241.2",      "Sh.NE_Kar-281.1",
            	"Sh.NE_Kar-37.2",       "Sh.NE_Lata-007.3",     "Sh.NE_Lata-033.1", 
            	"Sh.NE_Lata-078.1",     "Sh.NE_Lata-253.1",     "Sh.NE_Lata-275.2",
            	"Sh.NE_Lata-293.1",     "Sh.NE_Lata-294.1",     "Sh.NE_LibTB-009.2",
            	"Sh.NE_LibTB-010.1",    "Sh.NE_LibTB-022.1",    "Sh.NE_LibTB-028.1",    
            	"Sh.NE_LibTB-031.1",    "Sh.NE_NG-011.1",       "Sh.NE_NG-06.2", 
            	"Sh.NE_NG-089.1",       "Sh.NE_NG-236.1",       "Sh.NE_Seb-076.1",
            	"Sh.NE_Seb-078.2",      "Sh.NE_Seb-081.2",      "Sh.NE_Tiag-272.1",
            	"Sh.NE_YK-029.2",       "Sh.NE_YK-069.1",       "Sh.NE_YK-099.2",
            	"Sh.NE_YK-248.2",       "Sh.NE_Youri-069.2",    "Sh.NE_Youri-091.3", 
            	"Sh.TZ_PEM0063.1",      "Sh.TZ_PEM0075.1",      "Sh.TZ_PEM0076.1",
            	"Sh.TZ_PEM0079.1",      "Sh.TZ_PEM0089.2",      "Sh.TZ_PEM0094.2",      
            	"Sh.TZ_PEM0099.2",      "Sh.TZ_PEM0103.1",      "Sh.TZ_PEM0104.1",
            	"Sh.TZ_PEM0106.2",      "Sh.TZ_PEM0108.1",      "Sh.TZ_PEM0110.1", 
            	"Sh.TZ_PEM0114.3",      "Sh.TZ_PEM0115.4",      "Sh.TZ_PEM0120.1",
            	"Sh.TZ_PEM0125.1",      "Sh.TZ_PEM0126.1",      "Sh.TZ_PEM0127.1",
            	"Sh.TZ_PEM0128.1",      "Sh.TZ_PEM0130.1",      "Sh.TZ_PEM0133.1",
            	"Sh.TZ_PEM0139.2",      "Sh.TZ_PEM0145.3",      "Sh.TZ_PEM0154.1", 
            	"Sh.TZ_PEM0157.3",      "Sh.TZ_PEM0166.1",      "Sh.TZ_PEM0171.1",
            	"Sh.TZ_UNG0006.1",      "Sh.TZ_UNG0038.1",      "Sh.TZ_UNG0076.1",
            	"Sh.TZ_UNG0077.1",      "Sh.TZ_UNG0078.1",      "Sh.TZ_UNG0087.2",
           	    "Sh.TZ_UNG0089.3",      "Sh.TZ_UNG0092.3",      "Sh.TZ_UNG0099.1", 
            	"Sh.TZ_UNG0102.1",      "Sh.TZ_UNG0111.1",      "Sh.TZ_UNG0117.1",
            	"Sh.TZ_UNG0121.1",      "Sh.TZ_UNG0125.3",      "Sh.TZ_UNG0127.1",
            	"Sh.TZ_UNG0129.2",      "Sh.TZ_UNG0134.1",      "Sh.TZ_UNG0137.3",
            	"Sh.TZ_UNG0139.1",      "Sh.TZ_UNG0142.2",      "Sh.TZ_UNG0146.1"]

HAE_SRA_SAMPLES = ["ERR084970", "ERR037800", "SRR433865"]
MAN_SAMPLES     = ["Sm.BR_1278.1", "Sm.BR_0447.1", "Sm.BR_2039.1"]
BOV_SAMPLES     = ["ERR119622", "ERR103048"]
CUR_SAMPLES     = ["ERR119623"]

localrules: all
localrules: create_list_of_contigs
localrules: bqsr_r1_prep_gdbimport
localrules: bqsr_r2_prep_gdbimport

configfile: CONFIG_DIR + "config.yaml"

shell.prefix("source ~/.bash_profile; ")

rule all:
    input: 
        #xpand(MAP_DIR + "{sample}_processed.bam", sample=SAMPLES),
        #expand(R1_TABLE + "{sample}_recalibration_plot_r1.pdf", sample=SAMPLES),
        #BQSR_R1_DIR + "all_recalibration_plot_r1.pdf",
        #expand(R2_TABLE + "{sample}_recalibration_plot_r2.pdf", sample=SAMPLES),  
        #BQSR_R2_DIR + "all_recalibration_plot_r2.pdf",
        #REFPAN_DIR + "sH_ref_panel_snps.list",
        expand(OUT_MAP + "{sample}_vs_{ref}_processed.bam", sample=CUR_SAMPLES, ref="schCur"),
        expand(OUT_MAP + "{sample}_vs_{ref}_noDupes.metrics", sample=CUR_SAMPLES, ref="schCur"),
        expand(OUT_MAP + "{sample}_vs_{ref}_processed.bam.bai", sample=CUR_SAMPLES, ref="schCur")
       
#include: "scripts/01-filter_reads.snake"
#include: "scripts/02-base_recalibration_and_genotype_sh.snake"
#include: "scripts/03-build_sh_snp_refpanel.snake"
include: "scripts/04-genotype_outgroups.snake"

