# File:     05-pop_stats.snake
# Name:     Neal Platt
# Date:     13 April 2018
# Desc:     This Snake file is used to generate sH vcfs with sM coordinates and
#           calculate basic popgen stats including struture plots, pca, fst,etc.

# Reccomended
# Usage:    snakemake \
#               --printshellcmds \
#               --use-conda \
#               --cluster 'qsub -V -cwd -S /bin/bash -pe mpi {threads} -o {log}.log -e {log}.stderr' \
#               --jobs 100 \
#               --latency-wait 200 \
#               --snake \
#               --keep-going
#               --restart-times 2

# Reqs:     All required software should be installed/available through the 
#           singularity image, packaged conda env, or in the scripts dir.

# Comments: This portion of the workflow exectures after refining the genotypes

#-------------------------------------------------------------------------------
# To Do:
#   3] update comments
#   4] add cleanup on success
#   5] MITO ml tree
#   6] auto ml tree
#   7] networks (haplotypes?)
#   8] skyline plots
#
#-------------------------------------------------------------------------------

#setting up snakemake 
configfile: "config.yaml"
HOME_DIR = "/master/nplatt/schisto_hybridization/"
CONTIG_CHUNKS=50
ASSEMBLY = "schHae_v1.fa"

#setting work directory
RESULTS_DIR = HOME_DIR + "results/"
SCRIPTS_DIR = HOME_DIR + "scripts/"
DATA_DIR = HOME_DIR + "data/"
LOG_DIR = RESULTS_DIR + "logs/"
GENOME_DIR = DATA_DIR + "genome/"
GENOME = GENOME_DIR + ASSEMBLY
SEQ_DIR = DATA_DIR + "sequence_data/"
FILTER_READS_DIR = RESULTS_DIR + "01-filtered_reads/"
MAP_DIR = RESULTS_DIR + "02-mapped_reads/" 

GATK_DIR = RESULTS_DIR + "03-gatk/" 
BQSR_R1_DIR = GATK_DIR + "r1/"
R1_HC = BQSR_R1_DIR + "hc/"
R1_DB = BQSR_R1_DIR + "db/"
R1_GENO = BQSR_R1_DIR + "geno/"
R1_MERGE = BQSR_R1_DIR + "merge/"
R1_FILTER = BQSR_R1_DIR + "filter/"
R1_BAM = BQSR_R1_DIR + "bams/"
R1_TABLE = BQSR_R1_DIR + "tables/"

BQSR_R2_DIR = GATK_DIR + "r2/"
R2_HC = BQSR_R2_DIR + "hc/"
R2_DB = BQSR_R2_DIR + "db/"
R2_GENO = BQSR_R2_DIR + "geno/"
R2_MERGE = BQSR_R2_DIR + "merge/"
R2_FILTER = BQSR_R2_DIR + "filter/"
R2_BAM = BQSR_R2_DIR + "bams/"
R2_TABLE = BQSR_R2_DIR + "tables/"

GENO_DIR = GATK_DIR + "filter_geno/"
 
POPSTATS_DIR = RESULTS_DIR + "04-popstats/"
PHYLO_DIR = RESULTS_DIR + "05-phylogenetics"
WGA_DIR = RESULTS_DIR + "06-wga/" 

#setting work directory
workdir: HOME_DIR

#setting up local rules to be run on the head node
localrules: all


shell.prefix("source ~/.bash_profile; ")

#onsuccess:
#    shell("rm -r " + R2_HC)

#-------------------------------------------------------------------------------
# initializing rule - these files need to be generated...initiates the pipeline
rule all:
    input:
        expand(POPSTATS_DIR + "sH_filtered_{chr_type}-variants_sManLift.vcf", chr_type=["auto", "sex", "mito", "all"])

#-------------------------------------------------------------------------------
rule convert_vcf_to_fasta_w_ambig:
    input: 
        GENOME_DIR + "Smansoni_v7.fa"
    output:
        PHYLO_DIR + "sH_filtered_{chr_type}-SNPs_sManLift.fast"
    threads:
        1
    log:
        LOG_DIR + "index_sman_genome"
    shell:
        """
        #select SNPS
        python {SCRIPTS_DIR}vcf_to_diploid_fasta.py test_snps.vcf test_snps.fas
        """
##impute with beagle
#beagle \
#    gt=sHaem_mito_filtered.vcf \
#    impute=false \
#    out=sH_imputed_phased
##still imputes data

##vcf2fasta
#samtools faidx \
#    ../../data/genome/schHae_v1.fa AMPZ01026399.1 \
#    >schHae_v1_mt.fa 

##make haploid
#zcat sH_imputed_phased.vcf.gz | sed 's/|[0|1]//g' >sH_haplotypes.vcf

#vcf2fasta \
#    --reference schHae_v1_mt.fa  \
#    --prefix mito_ \
#    --default-ploidy 1 \
#    sH_haplotypes.vcf 

##change fasta headers
#for i in $(ls mito_Sh.*.fasta); do
#    HEADER=">"$(echo $i | cut -f2,3 -d"_")"#AMPZ01026399.1"
#    
#    sed -i "1 s/>.*/$HEADER/" $i
#
#done

#cat mito_Sh*.fasta >sH_mitochondria.fasta
#cat schHae_v1_mt.fa >>sH_mitochondria.fasta

#rm mito_Sh.*.fasta

#exclude uninformative in paup
#use seqmagick to convert to nexus
seqmagick convert test_snps.fasta test_snps.nex --alphabet dna

#in bin wget http://phylosolutions.com/paup-test/paup4a161_centos64.gz
begin paup;
export file=inf.nex format=nexus linebreaks=unix charsperline=all replace=yes;
end;

seqmagick convert inf.nex inf.fasta
#raxml

raxmlHPC -m GTRCAT -V -p 12345 -s test_snps.fas -n test_snps_infer_ML -# 100
raxmlHPC -m GTRCAT -V -p 12345 -b 98765 -s test_snps.fas -n test_snps_boot -# 1000
raxmlHPC -m GTRCAT -V -f b -t RAxML_bestTree.test_snps_infer_ML -z RAxML_bootstrap.test_snps_boot -n test_snps_bootFinal

#concurently LOCALLY in popart
seqmagick convert inf.nex inf.fasta

#running popart locally










#############################
#mito pca



